File: tests/csharp/JsonValue.cs
using CsCheck;
using FluentAssertions;
using System.ComponentModel.Design;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Nodes;
using Xunit;

namespace common.tests;

public class JsonValueTests
{
    [Fact]
    public void AsString_fails_if_the_node_is_null()
    {
        var value = (JsonValue?)null;
        var result = value.AsString();
        result.Should().BeError();
    }

    [Fact]
    public void AsString_fails_if_the_value_is_not_a_string()
    {
        var generator = from value in Generator.JsonValue
                        where value.GetValueKind() != JsonValueKind.String
                        select value;

        generator.Sample(value =>
        {
            var result = value.AsString();
            result.Should().BeError();
        });
    }

    [Fact]
    public void AsString_succeeds_if_the_value_is_a_string()
    {
        var generator = from value in Generator.JsonValue
                        where value.GetValueKind() == JsonValueKind.String
                        select value;
        generator.Sample(value =>
        {
            var result = value.AsString();
            result.Should().BeSuccess();
        });
    }

    [Fact]
    public void AsInt_fails_if_the_node_is_null()
    {
        var value = (JsonValue?)null;
        var result = value.AsInt();
        result.Should().BeError();
    }

    [Fact]
    public void AsInt_fails_if_the_value_is_not_an_int()
    {
        var generator = Gen.OneOf(from value in Gen.String
                                  select JsonValue.Create(value),
                                  from value in Gen.Bool
                                  select JsonValue.Create(value),
                                  from value in Gen.Char
                                  select JsonValue.Create(value),
                                  from value in Gen.Decimal
                                  where decimal.IsInteger(value) is false
                                  select JsonValue.Create(value),
                                  from value in Gen.Double
                                  where double.IsNaN(value) is false
                                  where double.IsInfinity(value) is false
                                  where double.IsInteger(value) is false
                                  select JsonValue.Create(value));

        generator.Sample(value =>
        {
            var result = value.AsInt();
            result.Should().BeError();
        });
    }

    [Fact]
    public void AsInt_succeeds_if_the_value_is_an_int()
    {
        var generator = from value in Gen.Int
                        select JsonValue.Create(value);

        generator.Sample(value =>
        {
            var result = value.AsInt();
            result.Should().BeSuccess();
        });
    }
}

File: tests/fsharp/JsonValue.fs
module JsonValueTests

open Xunit
open common
open Faqt


[<Fact>]
let ``asString with a non string value fails`` () =
    let gen = Gen.JsonValue.nonString

    Check.fromGen gen (fun value ->
        let result = JsonValue.asString value
        result.Should().BeFailure())

[<Fact>]
let ``asString with a string value succeeds`` () =
    let gen = Gen.JsonValue.string

    Check.fromGen gen (fun value ->
        let result = JsonValue.asString value
        result.Should().BeSuccess())

[<Fact>]
let ``asInt with a non integer value fails`` () =
    let gen = Gen.JsonValue.nonInteger

    Check.fromGen gen (fun value ->
        let result = JsonValue.asInt value
        result.Should().BeFailure())

[<Fact>]
let ``asInt with an integer value succeeds`` () =
    let gen = Gen.JsonValue.integer

    Check.fromGen gen (fun value ->
        let result = JsonValue.asInt value
        result.Should().BeSuccess())

[<Fact>]
let ``asAbsoluteUri with a non URI value fails`` () =
    let gen = Gen.JsonValue.nonAbsoluteUri

    Check.fromGen gen (fun value ->
        let result = JsonValue.asAbsoluteUri value
        result.Should().BeFailure())

[<Fact>]
let ``asAbsoluteUri with a URI value succeeds`` () =
    let gen = Gen.JsonValue.absoluteUri

    Check.fromGen gen (fun value ->
        let result = JsonValue.asAbsoluteUri value
        result.Should().BeSuccess())

[<Fact>]
let ``asGuid with a non GUID value fails`` () =
    let gen = Gen.JsonValue.nonGuid

    Check.fromGen gen (fun value ->
        let result = JsonValue.asGuid value
        result.Should().BeFailure())

[<Fact>]
let ``asGuid with a GUID value succeeds`` () =
    let gen = Gen.JsonValue.guid

    Check.fromGen gen (fun value ->
        let result = JsonValue.asGuid value
        result.Should().BeSuccess())

[<Fact>]
let ``asBool with a non boolean value fails`` () =
    let gen = Gen.JsonValue.nonBool

    Check.fromGen gen (fun value ->
        let result = JsonValue.asBool value
        result.Should().BeFailure())

[<Fact>]
let ``asBool with a boolean value succeeds`` () =
    let gen = Gen.JsonValue.bool

    Check.fromGen gen (fun value ->
        let result = JsonValue.asBool value
        result.Should().BeSuccess())
